Add Upgrade Guide notes for various deprecations and removals (#23068)
gradle/gradle: subprojects/docs/src/docs/userguide/migration/upgrading_version_7.adoc

Use the `mainClass` property instead.

==== AbstractExecTask API Cleanup

The deprecated `execResult` getter property of the `AbstractExecTask` task type has been removed.
Use the `executionResult` getter property instead.

==== AbstractTestTask API Cleanup

The deprecated `binResultsDir` property of the `AbstractTestTask` task type has been removed.
Use the `binaryResultsDirectory` property instead.

==== SourceDirectorySet API Cleanup

The deprecated `outputDir` property of the `SourceDirectorySet` type has been removed.
Use the `destinationDirectory` property instead.

==== VersionCatalog API Cleanup

The deprecated `findDependency(String)` method and `dependencyAliases` property of the `VersionCatalog` type have been removed.
Use the `findLibrary(String)` method and `libraryAliases` property instead.

The deprecated `alias(String)` method of the `VersionCatalogBuilder` type has been removed.
Use the `library(String, String, String)` or `plugin(String, String)` methods instead.

==== WorkerExecutor API Cleanup

The deprecated `submit(Class, Action)` method of the `WorkerExecutor` interface has been removed.

> Instead, obtain a `WorkQueue` via the `noIsolation()`. `classLoaderIsolation()`, `processIsolation()`, methods and use the `submit(Class, Action)` method on the `WorkQueue` instead.

Reviewer: ```suggestion
Instead, obtain a `WorkQueue` via the `noIsolation()`, `classLoaderIsolation()`, and `processIsolation()`, methods and use the `submit(Class, Action)` method on the `WorkQueue` instead.
```

Author: Thanks for taking such a careful look at this.

No changes are required for projects that already apply the
`java-base` plugin directly or indirectly through the `java`, `application`, `java-library`, or other JVM ecosystem plugins.

==== Upload Task should not be used

The `Upload` task remains deprecated and is now scheduled for removal in Gradle 9.0.
Although this type remains, it is no longer functional and will throw an exception upon running.
It is preserved solely to avoid breaking plugins.
Use the tasks in the `maven-publish` or `ivy-publish` plugins instead.

==== Configurations no longer allowed as Dependencies

Adding a Configuration as a dependency in the `dependencies` DSL block, or programmatically using the `DependencyHandler` classes' `doAdd(Configuration, Object, Closure)` method, is no longer allowed and will fail with an exception.

> You can extend configurations using the `extendsFrom(Configuration)` method on `Configuration` instead to replicate many of the results of this behavior.

Reviewer: ```suggestion
To replicate many aspects of this behavior, extend configurations using the `extendsFrom(Configuration)` method on `Configuration` instead.
```
